### PR TITLE
Update vulnerabilities to use new Tool specification 

### PR DESCRIPTION
**Pull request by shaikhu**
According to the [docs](https://cyclonedx.org/docs/1.5/json), since v1.5 the `Tools` format has been deprecated and a new format recommended. While the entity `Metada` has been updated to use the new format (new method `setToolChoice`), `Vulnerability` has not. This PR adds a similar method to `Vulnerability`.

### CODE REVIEW

File: src/main/java/org/cyclonedx/model/vulnerability/Vulnerability.java
Comment (nscuro) [inline src/main/java/org/cyclonedx/model/vulnerability/Vulnerability.java:97]: `tools` should be marked as deprecated to avoid confusion, similar to how it's done in `Metadata`:

https://github.com/CycloneDX/cyclonedx-core-java/blob/3702283abd718595c069025762979fb9dfcee26b/src/main/java/org/cyclonedx/model/Metadata.java#L63-L65

File: src/main/java/org/cyclonedx/model/vulnerability/Vulnerability.java
Comment (nscuro) [inline src/main/java/org/cyclonedx/model/vulnerability/Vulnerability.java:1]: `Metadata` has a custom deserializer to deal with the tool deserialization: https://github.com/CycloneDX/cyclonedx-core-java/blob/3702283abd718595c069025762979fb9dfcee26b/src/main/java/org/cyclonedx/util/deserializer/MetadataDeserializer.java#L146-L167

That doesn't seem to be the case for `Vulnerability`. @mr-zepol am I seeing this right that in order to support `ToolInformation` in `Vulnerability`, we need a custom deserializer there as well?

**Review by nscuro — CHANGES_REQUESTED** [review]

File: src/main/java/org/cyclonedx/model/vulnerability/Vulnerability.java
Comment (shaikhu) [reply src/main/java/org/cyclonedx/model/vulnerability/Vulnerability.java]: > `tools` should be marked as deprecated to avoid confusion, similar to how it's done in `Metadata`:
> 
> https://github.com/CycloneDX/cyclonedx-core-java/blob/3702283abd718595c069025762979fb9dfcee26b/src/main/java/org/cyclonedx/model/Metadata.java#L63-L65

👍 7fc564a0

File: src/main/java/org/cyclonedx/model/vulnerability/Vulnerability.java
Comment (shaikhu) [reply src/main/java/org/cyclonedx/model/vulnerability/Vulnerability.java:1]: > `Metadata` has a custom deserializer to deal with the tool deserialization:
> 
> https://github.com/CycloneDX/cyclonedx-core-java/blob/3702283abd718595c069025762979fb9dfcee26b/src/main/java/org/cyclonedx/util/deserializer/MetadataDeserializer.java#L146-L167
> 
> That doesn't seem to be the case for `Vulnerability`. @mr-zepol am I seeing this right that in order to support `ToolInformation` in `Vulnerability`, we need a custom deserializer there as well?

I think you are correct, a custom deserializer is needed. Updated the `VulnerabililtyDeserializer` with c55e9e66 and 5f563c33

File: src/main/java/org/cyclonedx/util/TimestampUtils.java
Comment (shaikhu) [inline src/main/java/org/cyclonedx/util/TimestampUtils.java:28]: I had to add this extra pattern to avoid deserialization errors. One example can be found [here](https://github.com/shaikhu/cyclonedx-core-java/blob/master/src/test/resources/1.4/valid-vulnerability-1.4.json#L63..L65). 

I don't know if this alt date format was intended or not, but rather then editing all the input files, I just added this pattern as a fallback if the default format failed.

File: src/main/java/org/cyclonedx/model/vulnerability/Vulnerability.java
Comment (mr-zepol) [reply src/main/java/org/cyclonedx/model/vulnerability/Vulnerability.java:1]: > `Metadata` has a custom deserializer to deal with the tool deserialization:
> 
> https://github.com/CycloneDX/cyclonedx-core-java/blob/3702283abd718595c069025762979fb9dfcee26b/src/main/java/org/cyclonedx/util/deserializer/MetadataDeserializer.java#L146-L167
> 
> That doesn't seem to be the case for `Vulnerability`. @mr-zepol am I seeing this right that in order to support `ToolInformation` in `Vulnerability`, we need a custom deserializer there as well?

yes, I was trying to check with annotations but it's hard to achieve given both XML and JSON

File: src/main/java/org/cyclonedx/util/TimestampUtils.java
Comment (mr-zepol) [reply src/main/java/org/cyclonedx/util/TimestampUtils.java:28]: I don't think we need to worry about dates here, I get your point, but maybe would make more sense to update the test files if they are different from the other, just to reduce complexity of this change

File: src/main/java/org/cyclonedx/util/deserializer/AffectDeserializer.java
Comment (mr-zepol) [inline src/main/java/org/cyclonedx/util/deserializer/AffectDeserializer.java:34]: JOOC Why is this other serializer needed? is not done automatically with the VulnerabilitySerializer?

Comment (mr-zepol) [reply src/main/java/org/cyclonedx/util/deserializer/AffectDeserializer.java:34]: Same question for the list

File: src/main/java/org/cyclonedx/util/deserializer/ToolDeserializer.java
Comment (mr-zepol) [inline src/main/java/org/cyclonedx/util/deserializer/ToolDeserializer.java:40]: without this does not parse the tool? I don't see any logic

File: src/test/java/org/cyclonedx/parsers/AbstractParserTest.java
Comment (mr-zepol) [inline src/test/java/org/cyclonedx/parsers/AbstractParserTest.java:564]: this seems the same as the condition above, so I would merge both

File: src/main/java/org/cyclonedx/util/deserializer/VulnerabilityDeserializer.java
Comment (mr-zepol) [inline src/main/java/org/cyclonedx/util/deserializer/VulnerabilityDeserializer.java:284]: I think we have something similar in other deserializer

File: src/test/java/org/cyclonedx/parsers/AbstractParserTest.java
Comment (mr-zepol) [inline src/test/java/org/cyclonedx/parsers/AbstractParserTest.java:552]: Other good tests that would check if this is working as expected, is loading the JSON and converting to a XML and the other way around, BomXMLGeneratorTest and BomJsonGeneratorTest you can check some examples there

File: src/main/java/org/cyclonedx/util/deserializer/VulnerabilityDeserializer.java
Comment (mr-zepol) [inline src/main/java/org/cyclonedx/util/deserializer/VulnerabilityDeserializer.java:213]: what if it's an older version and has only the toolChoice?

File: src/test/java/org/cyclonedx/parsers/AbstractParserTest.java
Comment (shaikhu) [reply src/test/java/org/cyclonedx/parsers/AbstractParserTest.java:551]: > this seems the same as the condition above, so I would merge both

👍 21abe5f0

File: src/main/java/org/cyclonedx/util/TimestampUtils.java
Comment (shaikhu) [reply src/main/java/org/cyclonedx/util/TimestampUtils.java]: > I don't think we need to worry about dates here, I get your point, but maybe would make more sense to update the test files if they are different from the other, just to reduce complexity of this change

Updated. fa67e9d1

File: src/test/java/org/cyclonedx/parsers/AbstractParserTest.java
Comment (shaikhu) [reply src/test/java/org/cyclonedx/parsers/AbstractParserTest.java]: > Other good tests that would check if this is working as expected, is loading the JSON and converting to a XML and the other way around, BomXMLGeneratorTest and BomJsonGeneratorTest you can check some examples there

Great idea 👍 

IIUC, this is the [test ](https://github.com/shaikhu/cyclonedx-core-java/blob/409aa99c442bdfd5e8c8e71b5f18b7e22072e9e4/src/test/java/org/cyclonedx/BomXmlGeneratorTest.java#L183) for `BomXMLGeneratorTest` and this is the [test ](https://github.com/shaikhu/cyclonedx-core-java/blob/409aa99c442bdfd5e8c8e71b5f18b7e22072e9e4/src/test/java/org/cyclonedx/BomJsonGeneratorTest.java#L155) for `BomJsonGeneratorTest`. 

For the files `bom-1.5.xml` and `bom-1.5.json` (which are included in the above tests) the new `Tool` specification is used. See [here ](https://github.com/shaikhu/cyclonedx-core-java/blob/409aa99c442bdfd5e8c8e71b5f18b7e22072e9e4/src/test/resources/bom-1.5.json#L575) and [here](https://github.com/shaikhu/cyclonedx-core-java/blob/409aa99c442bdfd5e8c8e71b5f18b7e22072e9e4/src/test/resources/bom-1.5.xml#L441). 

I have updated the input test files for `/1.6/valid-bom-1.6.xml|json`  as for some reason they where using the deprecated `Tool` specification, which seems incorrect imo. See 409aa99c

File: src/main/java/org/cyclonedx/util/deserializer/VulnerabilityDeserializer.java
Comment (shaikhu) [reply src/main/java/org/cyclonedx/util/deserializer/VulnerabilityDeserializer.java:213]: > what if it's an older version and has only the toolChoice?

Looking at the [docs](https://cyclonedx.org/docs/1.6/json/#vulnerabilities_items_tools), both the new and deprecated field has the same name `Tools`. Also the [xsd](https://github.com/CycloneDX/cyclonedx-core-java/blob/dde00242ae0271e38a0e4312fad0136d801f219c/src/main/resources/bom-1.6.xsd#L173) only defines `Tools` not `Tools` and `ToolChoice`. 

However looking at the docs, the field is of type `Object`, while the previous type was `Array`. For this reason I'm dropping the `JacksonXmlElementWrapper` annotation on the new type.  See 691649c4. Let me know if you think it should be added back.

File: src/main/java/org/cyclonedx/util/deserializer/AffectDeserializer.java
Comment (shaikhu) [reply src/main/java/org/cyclonedx/util/deserializer/AffectDeserializer.java:34]: > JOOC Why is this other serializer needed? is not done automatically with the VulnerabilitySerializer?

It wasn't done automatically. Without the affects deserializer the following [test ](https://github.com/shaikhu/cyclonedx-core-java/blob/691649c4d0f7adeba4a27ddb1ddf33a69f0c9c1b/src/test/java/org/cyclonedx/parsers/AbstractParserTest.java#L597)failed.

Comment (shaikhu) [reply src/main/java/org/cyclonedx/util/deserializer/AffectDeserializer.java:34]: > Same question for the list

See above comment.

File: src/main/java/org/cyclonedx/util/deserializer/ToolDeserializer.java
Comment (shaikhu) [reply src/main/java/org/cyclonedx/util/deserializer/ToolDeserializer.java:40]: > without this does not parse the tool? 
Nope. Throws a null pointer.

> I don't see any logic
I think the mapper is smart enough to use the class definition to how to deserialize? Similar to [here](https://github.com/shaikhu/cyclonedx-core-java/blob/691649c4d0f7adeba4a27ddb1ddf33a69f0c9c1b/src/main/java/org/cyclonedx/util/deserializer/SignatoryDeserializer.java#L74..L75).

File: src/main/java/org/cyclonedx/util/deserializer/VulnerabilityDeserializer.java
Comment (shaikhu) [reply src/main/java/org/cyclonedx/util/deserializer/VulnerabilityDeserializer.java:284]: > I think we have something similar in other deserializer

Where? 😅. I've looked but the only thing I could find was this [code](https://github.com/shaikhu/cyclonedx-core-java/blob/master/src/main/java/org/cyclonedx/util/deserializer/SignatoryDeserializer.java#L69..L78), which converts a single Organization, where the code here is converting a list of organizations so is a little more complex.

File: src/main/java/org/cyclonedx/util/deserializer/VulnerabilityDeserializer.java
Comment (mr-zepol) [reply src/main/java/org/cyclonedx/util/deserializer/VulnerabilityDeserializer.java:213]: Get it, this if looks the same as the one in the metadata deserializer, I would have one method for both in a util classes or something

File: src/main/java/org/cyclonedx/util/deserializer/VulnerabilityDeserializer.java
Comment (mr-zepol) [reply src/main/java/org/cyclonedx/util/deserializer/VulnerabilityDeserializer.java:284]: you are right, this was removed with some changes

**Comment by mr-zepol** [comment]
@shaikhu new fields were introduced to the vulnerability class, so you might need to rebase and add them to the serializer you are creating

File: src/main/java/org/cyclonedx/util/deserializer/VulnerabilityDeserializer.java
Comment (shaikhu) [reply src/main/java/org/cyclonedx/util/deserializer/VulnerabilityDeserializer.java:210]: > Get it, this if looks the same as the one in the metadata deserializer, I would have one method for both in a util classes or something

28e15c1

**Comment by shaikhu** [comment]
> @shaikhu new fields were introduced to the vulnerability class, so you might need to rebase and add them to the serializer you are creating

978fccaa

File: src/main/java/org/cyclonedx/util/TimestampUtils.java
Comment (shaikhu) [reply src/main/java/org/cyclonedx/util/TimestampUtils.java:27]: > I would rather not introduce this for some serializers only, 

So this was already there for `MetadataDeserializer` [here](https://github.com/CycloneDX/cyclonedx-core-java/blob/eb0732750553b5bc61472131e5db89e29591ead4/src/main/java/org/cyclonedx/util/deserializer/MetadataDeserializer.java#L32). Rather than duplicating the `SimpleDateFormat` in `VulnerabilityDeserializer` I thought I would put it inside a util class. Basically similar to this [suggestion](https://github.com/CycloneDX/cyclonedx-core-java/pull/438#discussion_r1757489839).

**Review by mr-zepol — APPROVED** [review]
+1 LGTM, I can run some tests after is merged and if I find something I can check with @shaikhu

**Comment by mr-zepol** [comment]
@nscuro are we okay to merge this one?

**Review by nscuro — APPROVED** [review]

**Pull request merged**
